Support looking through lights, round 2

danieldresser opened the pull request:
Started trying out John's suggestion on #1830 

Looks like it's basically working.  The LightToCamera node adds camera objects on lights, and adjusts transforms to get the projections right ( still need to set clipping planes properly ).

Next step is how would this actually connect up to the scene view?  We wouldn't want to actually render this scene with extra cameras added to it and transforms adjusted, so we would need to merge the camera we want to render with back into the scene.

One possibility:  use a Group node to group the whole scene into a __cameraRoot group, set it's visibility to false, and then parent it back to /.  Then the scene view just needs to prepend /__cameraRoot on the front of the camera path, and it will pick up an accurate camera if you give it a light.

Possible issues with this approach:
* Requires support for rendering from invisible camera ( currently works fine with GL view, and I think we decided that the long term convention was going to be that rendering from invisible cameras is OK? )
* Andrew thinks it's horrifying to have an extra copy of the scene under __cameraRoot in an internal processing network in the scene view.  I'm not sure if it's horrifying - I feel like relying on lazy evaluation to mean that we will just pick out the camera path we're interested in without evaluating extra stuff might be a gaffery thing to do?  It would be completely invisible to users anyway.
* When looking through a light, you still want to see the light viewport visualiser ( so you can see penumbra angle at least, and more stuff on the IE visualisers ), so I would have to modify the way that SceneView::LookThrough::updateLookThroughCamera() sets m_view->hideFilter().  I'm not quite sure why this was needed in the first place, the camera visualiser doesn't seem to show up when the view location is right at its origin anyway.  Maybe I need to treat objects in the __lights set differently?

Or maybe you've got a way better idea how this is all supposed to work?  Andrew suggested adding cameras as children of the lights instead of replacing them - I'm really hoping that if I need to do hierarchy manipulation, it will be doable using existing Gaffer nodes rather than implementing custom hierarchy manipulation.

johnhaddon commented:
Perhaps I've missed something, but I think this can be a lot simpler. The `SceneView::LookThrough` class already builds a [small internal processing network](https://github.com/ImageEngine/gaffer/blob/master/src/GafferSceneUI/SceneView.cpp#L784) to remove motion blur from the main input scene before retrieving the camera. This is internal to the LookThrough, so doesn't affect what is seen by the SceneGadget used to actually do the GL drawing. I assumed we could also just insert a LightsToCamera node into that internal network and then continue as normal - because it's only used internally we should just be able to use a SetFilter to convert all the lights unconditionally in this way. Any reason that won't work?

As for hiding the camera, I feel like it must have been necessary at some point, otherwise I wouldn't have gone to the trouble. Perhaps the resolution gate overlay hides it at the moment because the two happen to line up exactly? Hiding conditionally based on light/camera type doesn't seem too hard though?

danieldresser commented:
> The SceneView::LookThrough class already builds a small internal processing

Ah, I should have spotted that.  Yep, that makes everything easy.

> Hiding conditionally based on light/camera type doesn't seem too hard though?

Yeah, should be pretty straightforward.  What do you think the logic should be?
Only hide it if m_lookThroughCamera is in the __cameras set of the unmodified scene?

danieldresser commented:
Going to do some more testing tomorrow, but as far as I can see, this is now working.

danieldresser commented:
Note to self - still need to set clip planes.

johnhaddon commented:
I'll do a review once you're done, but I wanted to ask what the plan is for the picking of lights in the look-through field at the top of the viewer? At the moment you can browse for cameras, can we add a checkbox there to also show lights?

danieldresser commented:
OK, now set clip planes appropriately ( though looking through a distant light makes more sense with https://github.com/ImageEngine/cortex/pull/522 ).

As for "can we add a checkbox there to also show lights?", I probably should have just left this up to you, but since you asked I had a go.  The 3rd commit does what I can using CompoundPathFilter : a checkbox for "Show only cameras", and a second one for "Show only cameras and lights", both of which default on, matching previous behaviour.  You can turn off the first to get both. Maybe it would be better to add a union mode to CompoundPathFilter, but if you think that's called for, maybe you could merge the first two commits and worry about that later?

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+size_t LightToCamera::g_firstPlugIndex = 0;
+
+LightToCamera::LightToCamera( const std::string &name )
+	:	SceneElementProcessor( name )
```
Prefer `SceneElementProcessor( name, GafferScene::Filter::NoMatch )`, to create a node which requires a filter before it does anything. We have a todo item to do this for all existing nodes, so in the meantime we shouldn't add any more with the wrong behaviour.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+namespace 
+{
+
+	// TODO: copied from src/GafferSceneUI/StandardLightVisualiser.cpp
```
Perhaps we should make this todo about introducing the LightAccessor class?

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+	// Should this live somewhere central? I think John talked about using the same functionality
+	// to drive light manipulators? 
+	template<typename T>
+	T parameter( InternedString metadataTarget, const IECore::CompoundData *parameters, InternedString parameterNameMetadata, T defaultValue )
```
No need for indentation at namespace level.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	// If the attributes contain a light, return a corresponding camera
+	// Just returns the first light it finds, which could produce inconsistent results if you
```
I don't mind this as a strategy for now, but I don't really buy this as a justification. If someone were making  a multi-renderer setup without the fancy adaptors we have, they might very much want to have light shaders for more than one renderer assigned at the same location.

danieldresser replied:
Hopefully if the two lights are for a multirenderer setup of the same scene, they won't have dramatically differing cone angles, and it won't matter too much which you look through.  More problematic for shadow maps or something, but if just about viewport preview it seems fairly OK.  Can remove last sentence of comment.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+			float screenWindowScale = 1.0f;
+			if( type->readable() == "distant" )
+			{
+			    const float locatorScale = parameter<float>( metadataTarget, shaderParameters, "locatorScaleParameter", 1 );
```
Indentation is out of whack.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+				}
+
+				float outerAngle = coneAngle;
+				if( penumbraType && *penumbraType == "outset" )
```
Do we also need to think about the "absolute" penumbra type?

danieldresser replied:
Currently, I match the fov to the outer angle, and both the inset and absolute types set outerAngle = coneAngle.  Can add comment to this effect.

johnhaddon replied:
Cool - thanks for clarifying.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+
+					M44f transformOffset;
+					transformOffset.setTranslation( V3f( 0.0f, 0.0f, focalPointOffset ) );
+					result->parameters()["__projectionTransformOffset"] = new M44fData( transformOffset );
```
Exposing the internal implementation in a parameter like this is pretty inelegant. How about one of the alternatives?

1. At least stuff it in the blindData rather than the parameters.
2. Have this intermediate camera be computed on an internal plug, and clean it up on the way out of `computeObject()`.
3. Factor out the stuff to get outerAngle so that it can be reused in `computeTransform()`. This would be my preference, since I believe this is how we would approach things if we had a `LightAccessor`, and I think in the long run this will be the neatest.

danieldresser replied:
I feel bad about repeating all computation twice, but I guess it shouldn't be a bottleneck.  I think I ended up going this way partly because I couldn't think of a clean way to factor it out - most of the logic is going to have to be repeated.  Maybe just have:
void convertLightToCamera( IECore::ConstCompoundObjectPtr attributes, int resolution, IECore::CameraPtr &outCamera, M44f &projectionTransformOffset ),
and always ignore one of the two outputs?

johnhaddon replied:
It feels wrong to me to allocate a camera only to free it again when getting the transform. Couldn't things be refactored fairly sensibly like so?

```
// This would contain the loop over the attributes to find a light, removing an
// entire level of nesting from convertLightToCamera().
const Shader *light( const CompoundObject *attributes );
// These mimic methods we might expect to see on a LightAccessor
// class in the future.
std::string &type( const Shader *light );
float outerAngle( const Shader *light );
// This refactors out the shared logic we need for the transform.
float focalPointOffset( const Shader *light );
```

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+
+		return NULL;
+	}
+}
```
`} // namespace` - helps clarify things after removing indentation.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+
+IECore::ConstCompoundObjectPtr LightToCamera::computeProcessedAttributes( const ScenePath &path, const Gaffer::Context *context, IECore::ConstCompoundObjectPtr inputAttributes ) const
+{
+	// Remove any light attributes - because we may be shifting the transform, they could be invalidated,
```
I don't think this needs to be the justication - we're converting lights to cameras, so the lights should be deleted regardless of any shifting transforms.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -0,0 +1,312 @@
+	addChild( new IntPlug( "resolution", Plug::In, 512 ) );
+
+	// Fast pass-throughs for things we don't modify
+	outPlug()->boundPlug()->setInput( inPlug()->boundPlug() );
```
We need to disconnect `setNamesPlug()` and `setPlug()` here too, so we can compute the updates `__lights` and `__cameras` sets.

johnhaddon commented on src/GafferSceneUI/SceneView.cpp:
```diff
@@ -781,14 +783,29 @@ class SceneView::LookThrough : public boost::signals::trackable
+			m_internalNodes.push_back( lightFilter );
+			m_internalNodes.push_back( lightConverter );
+
+			
```
Could you limit blank lines to one at a time? Splitting things for readability is great, but I'd like to balance it with pushing things off screen vertically.

johnhaddon commented:
I'm getting a bunch of GL errors when looking through `/group/light` in this simple setup.

```
import Gaffer
import GafferArnold
import GafferScene
import IECore

__children = {}

__children["Cube"] = GafferScene.Cube( "Cube" )
parent.addChild( __children["Cube"] )
__children["Cube"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Cube"]["__uiPosition"].setValue( IECore.V2f( -12.1666975, 3.77459931 ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in3", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["__uiPosition"].setValue( IECore.V2f( 2.33330107, -5.38949442 ) )
__children["Camera"] = GafferScene.Camera( "Camera" )
parent.addChild( __children["Camera"] )
__children["Camera"]["transform"]["translate"].setValue( IECore.V3f( 0, 0, 6 ) )
__children["Camera"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Camera"]["__uiPosition"].setValue( IECore.V2f( 0.833301306, 3.77626252 ) )
__children["spot_light"] = GafferArnold.ArnoldLight( "spot_light" )
parent.addChild( __children["spot_light"] )
__children["spot_light"]["transform"]["translate"].setValue( IECore.V3f( 0, 0, 6 ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "radius", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "lens_radius", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "cone_angle", defaultValue = 65.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "penumbra_angle", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "cosine_power", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "aspect_ratio", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.StringPlug( "decay_type", defaultValue = 'quadratic', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.Color3fPlug( "color", defaultValue = IECore.Color3f( 1, 1, 1 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "intensity", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "exposure", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.BoolPlug( "cast_shadows", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.BoolPlug( "cast_volumetric_shadows", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "shadow_density", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.Color3fPlug( "shadow_color", defaultValue = IECore.Color3f( 0, 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.IntPlug( "samples", defaultValue = 1, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.BoolPlug( "normalize", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.BoolPlug( "affect_diffuse", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.BoolPlug( "affect_specular", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.BoolPlug( "affect_volumetrics", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "diffuse", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "specular", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "sss", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.FloatPlug( "indirect", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.IntPlug( "max_bounces", defaultValue = 999, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["parameters"].addChild( Gaffer.IntPlug( "volume_samples", defaultValue = 2, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["__shaderName"].setValue( 'spot_light' )
__children["spot_light"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"]["__uiPosition"].setValue( IECore.V2f( 13.8333015, 3.77456856 ) )
__children["Group"]["in"]["in0"].setInput( __children["Cube"]["out"] )
__children["Group"]["in"]["in1"].setInput( __children["Camera"]["out"] )
__children["Group"]["in"]["in2"].setInput( __children["spot_light"]["out"] )


del __children
```

johnhaddon commented:
> As for "can we add a checkbox there to also show lights?", I probably should have just left this up to you, but since you asked I had a go.

It feels a bit awkward doesn't it? I wonder if it would be better to have a single "Show Lights as well" type of a checkbox, so it's just cameras, or cameras and lights, and never anything you can't look through?

danieldresser commented:
Good catch on that failure.  Hadn't taken care to test lights with 0 lens radius after adding support for lens radius lights.  Fixed now.

As for the toggle setup, just "Show Lights" sounds good.  How would implementing that work?  Add a hidden plug to the SceneView that outputs from a filter, and then the _LookThroughPlugValueWidget camera widget hooks up to plug.node()["__lightsAndCamerasOnly"], instead of plug.node()["in"]?

Starting cleaning up little things, but won't get to updating lights and camera sets til tomorrow.

johnhaddon commented:
For the toggle, maybe we can have one filter which is the union of cameras and lights (using SetFilter, UnionFilter and SceneFilterPathFilter), and one which is just the cameras. Then combine them with CompoundPathFilter, and configure the UI so that only the UI for the second filter is shown, with the checkbox disabling the filter? If you'd rather I took a look at that, then feel free to remove your UI commit.

One more thing I didn't mention yesterday - I'm not confident about the public resolution plug. I am confident that the default output from this thing should be a square screen window, but I'm not sure that a resolution plug would be the right interface for defining the form of that. Can we just default the resolution to square internally for now, and then let actual use cases dictate a public interface we add later? We always have the Parameters and StandardOptions nodes to allow folks to monkey with this stuff downstream if needs be in the meantime...

danieldresser commented:
OK, I think I've handled all the little ones, and squashed those changes in.

The bigger issues:

* UI for "Show Lights" - I've dropped this commit, sounds like you can much more easily figure out modifying CompoundPathFilter UIs than I can.
* Switching convertLightToCamera into something that looks more like a light accessors are going to look.  Not perfectly elegant, and I'm not sure how you'll feel about light() returning shaderParameters and a metadataTarget instead of a Shader object - but it does need to pass the name somehow.  Given that it will hopefully be rewritten fairly soon anyway when we do light accessors, hopefully this is good enough?
* Modifying sets correctly.  This feels far more complicated and scary than any other part of this code.  I was kind of hoping I could create a subnetwork to modify the sets using existing Gaffer set manipulation, but I couldn't think of a way to do this.  I ended up copying from the Prune node as an example of something where the filter affects the sets, and I don't feel entirely confident I won't have missed something like how hashSet needs to remove the path from the context.  This requires the modified acceptsInput - I wasn't sure of whether I should have aimed for something more like the set node instead, with the intermediate PathMatcherData plug.  I should probably write some test cases, though I don't know if I'm familiar enough with things to potentially mess up in Gaffer set processing to know the right tests to write.  Not sure about the performance implications of "outputSet.addPaths( lightSet.subTree( *pIt ), *pIt );" either. ￼ Maybe you've got some completely different idea of how I should have done this without duplicating some fairly complex logic?

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -49,12 +49,11 @@ using namespace Gaffer;
-namespace 
+/// \todo: This stuff should all end up as part of the light accessor API once that exists
+/// In the meantime, I've tried to set it up vaguely similarly to how that will work
+namespace GoingToBeLightAccessor
```
Let's just stick with an anonymous namespace, since LightAccessors will look different (be classes) and also we don't want any of these functions to be exported.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -73,11 +72,8 @@ T parameter( InternedString metadataTarget, const IECore::CompoundData *paramete
-// be avoided anyway
-IECore::CameraPtr convertLightToCamera( IECore::ConstCompoundObjectPtr attributes )
+// Return the first light shader found in attributes
+void light( const CompoundObject *attributes, const IECore::CompoundData* &shaderParameters, std::string &metadataTarget )
```
I had wondered about having this method return the parameters and name (rather than the Shader), but for a different reason. I was wondering if you would want to maintain support for old lights in caches where the light will be an `IECore::Light` rather than an `IECore::Shader`, and this interface would have allowed that. Is that worth adding?

I'm not particularly fussed because this is all internal implementation detail, but I'd probably use the return value for the shader parameters, because I find output arguments a bit harder to follow.

johnhaddon commented on include/GafferScene/LightToCamera.h:
```diff
@@ -78,6 +79,14 @@ class LightToCamera : public SceneElementProcessor
 		virtual Imath::M44f computeProcessedTransform( const ScenePath &path, const Gaffer::Context *context, const Imath::M44f &inputTransform ) const;
 
+		virtual void hashSetNames( const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const;
+        virtual void hashSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const;
```
Indentation.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -335,3 +385,139 @@ IECore::ConstCompoundObjectPtr LightToCamera::computeProcessedAttributes( const
+	}
+	else
+	{
+		ContextPtr tmpContext = new Context( *Context::current(), Context::Borrowed );
```
This can be replaced with `lightSetData = inPlug()->set( g_lightsSetName )`.

johnhaddon commented on src/GafferScene/LightToCamera.cpp:
```diff
@@ -335,3 +385,139 @@ IECore::ConstCompoundObjectPtr LightToCamera::computeProcessedAttributes( const
+	ContextPtr tmpContext = filterContext( context );
+	Context::Scope scopedContext( tmpContext.get() );
+
+	for( PathMatcher::RawIterator pIt = lightSet.begin(), peIt = lightSet.end(); pIt != peIt; )
```
This logic looks wrong to me, and more complicated than necessary. Here's an alternative stab at it :

```
for( PathMatcher::Iterator pIt = lightSet.begin(), peIt = lightSet.end(); pIt != peIt; ++pIt )
{
    const int m = filterPlug()->getValue();
    if( m & Filter::ExactMatch )
    {
        if( isLightSet )
        {
            outputSet.removePath( *pIt );
        }
        else
        {
            outputSet.addPath( *pIt );
        }
    }
}
```

How does that look?

johnhaddon commented:
> UI for "Show Lights" - I've dropped this commit, sounds like you can much more easily figure out modifying CompoundPathFilter UIs than I can.

Cool, no problem.

>  Switching convertLightToCamera into something that looks more like a light accessors are going to look. Not perfectly elegant, and I'm not sure how you'll feel about light() returning shaderParameters and a metadataTarget instead of a Shader object - but it does need to pass the name somehow. Given that it will hopefully be rewritten fairly soon anyway when we do light accessors, hopefully this is good enough?

Looks good to me - stops us leaking the internal details as camera parameters, and is more readable too, since the logic of getting lights and their properties is abstracted out from the code that generates transforms and cameras.

> Modifying sets correctly. This feels far more complicated and scary than any other part of this code. I was kind of hoping I could create a subnetwork to modify the sets using existing Gaffer set manipulation, but I couldn't think of a way to do this.

Nice suggestion - a subnetwork would be a great way of going about it. I've been thinking about adding a FilterResults node that would output all the matching paths for an input filter, and something like that might help here I think. My use case was to wedge a single task over a bunch of objects (render from all cameras for instance) so it would have had a string array output, but I think the same node could have a PathMatcherData output too, which might be useful for this case. One for another time though?

> Maybe you've got some completely different idea of how I should have done this without duplicating some fairly complex logic?

I don't have a better idea I'm afraid. But I think the scariest logic is the bit that I suggested a much simpler replacement for above, so if that looks OK to you, I'll be happy with that for now. Perhaps add a todo talking about how it would be nice to use an internal FilterResults node to simplify things instead? We might even be able to make that work for Isolate/Prune too, which would allow us to contain some of our ugliest logic all in one place.

> I should probably write some test cases, though I don't know if I'm familiar enough with things to potentially mess up in Gaffer set processing to know the right tests to write. 

Yes you should, before you started! Since we're backfilling though, let's just add a basic test for spot->persp and distant->ortho and a test with two lights with only one being converted to test the set output. Better to have something than nothing.

Couple of other small things :

- There's still a `resolutionPlug()` method in the header.
- I think we should always output a camera from `computeProcessedObject()`, even if we don't have dedicated conversion logic for whatever light we find. By modifying the cameras set we're kindof promising that there will be a camera there...

Seems like we're pretty much there, if you're happy with my set processing implementation.

johnhaddon commented:
One other small thing - when looking through a spot with a non-zero lens radius, I'm finding the colour swatch / exposure meter thing comes into view. Could it perhaps just be a case of adding a teensy offset on to the near clipping plane?

danieldresser commented:
OK, cleaned up the things you mentioned, and squashed it all down, hopefully that's about good to go.

> My use case was to wedge a single task over a bunch of objects (render from all cameras for instance) so it would have had a string array output, but I think the same node could have a PathMatcherData output too, which might be useful for this case. One for another time though?

Definitely one for another time.  To really make it convenient, I guess I would need the FilterResults node, then a node that could union or difference PathMatcherData, and then could I plug that straight to the hidden PatchMatcherData plug on a set node to make a set out of it again?

> Yes you should, before you started!

When I started, this was one very simple function in SceneAlgo.  Any tests I wrote for that wouldn't help me much now ...
I've added a few basic tests.

johnhaddon commented:
> OK, cleaned up the things you mentioned, and squashed it all down, hopefully that's about good to go.

Sweet - thanks. I've made a few minor tweaks and added the UI as we discussed, and made an updated PR in #1846. Hopefully if Travis gives me the nod I'll have that merged by start of day at your end. Thanks for all your work pushing the lights/visualisers stuff - it's a direction we really need to push in, so it's great to see the progress.

> Definitely one for another time. To really make it convenient, I guess I would need the FilterResults node, then a node that could union or difference PathMatcherData, and then could I plug that straight to the hidden PatchMatcherData plug on a set node to make a set out of it again?

The Set node already has Add/Remove operations, so I think the FilterResults would be the only additional thing we'd need. I'm quite tempted to try to sneak that in at some point, as I think it'd be really handy.

> When I started, this was one very simple function in SceneAlgo. Any tests I wrote for that wouldn't help me much now.

They would have caught the bug whereby adding lens radius support broke the look-through node (the one I found through manual testing), and given us a safety net when doing the refactoring.

johnhaddon commented:
Closing, after merging #1846.

andrewkaufman commented:
Carsten was interested in that `FilterResults` node, though I can't remember what his use case was (or if it was just theoretical)... but you might not have to "sneak" it if we can dig up his reasoning.
